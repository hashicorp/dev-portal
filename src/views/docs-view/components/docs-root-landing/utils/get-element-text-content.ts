/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { Children, ReactChild, ReactElement } from 'react'

/**
 * Given a React element,
 * Return the text content of that element.
 */
export function getElementTextContent(element: ReactElement) {
	let textContent = ''
	Children.forEach(element.props.children, (child: ReactChild) => {
		if (typeof child === 'string') {
			textContent += child
		}
	})
	return textContent
}
