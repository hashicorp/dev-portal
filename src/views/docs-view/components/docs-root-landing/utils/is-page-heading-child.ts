import { ReactChild } from 'react'

/**
 * Given a `ReactChild`, from MDX content,
 * Return `true` if the element is an `h1` element, or `false` otherwise.
 */
export function isPageHeadingChild(child: ReactChild): boolean {
	// Skip over string and number literals (we can't access props on these)
	if (typeof child === 'string' || typeof child === 'number') {
		return false
	}

	// Otherwise, we can check props to determine if this is the page heading
	return child.props.originalType === 'h1'
}
