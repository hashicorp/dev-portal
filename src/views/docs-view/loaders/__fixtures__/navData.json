[
	{ "title": "Overview", "path": "extending-waypoint" },
	{
		"title": "Creating a Waypoint Plugin",
		"routes": [
			{ "title": "Overview", "path": "extending-waypoint/creating-plugins" },
			{
				"title": "Registering Plugin Components",
				"path": "extending-waypoint/creating-plugins/main"
			},
			{
				"title": "Handling Configuration",
				"path": "extending-waypoint/creating-plugins/configuration"
			},
			{
				"title": "Implementing the Builder Interface",
				"path": "extending-waypoint/creating-plugins/build-interface"
			},
			{
				"title": "Compiling the Plugin",
				"path": "extending-waypoint/creating-plugins/compiling"
			},
			{
				"title": "Creating an Example Application",
				"path": "extending-waypoint/creating-plugins/example-application"
			},
			{
				"title": "Testing the Plugin",
				"path": "extending-waypoint/creating-plugins/testing"
			}
		]
	},
	{ "title": "Initializing the SDK", "path": "extending-waypoint/main-func" },
	{
		"title": "Passing Values Between Components",
		"path": "extending-waypoint/passing-values"
	},
	{
		"title": "Plugin Interfaces",
		"routes": [
			{ "title": "Overview", "path": "extending-waypoint/plugin-interfaces" },
			{
				"title": "Authenticator",
				"path": "extending-waypoint/plugin-interfaces/authenticator"
			},
			{
				"title": "Configurable",
				"path": "extending-waypoint/plugin-interfaces/configurable"
			},
			{
				"title": "ConfigurableNotify",
				"path": "extending-waypoint/plugin-interfaces/configurable-notify"
			},
			{
				"title": "Builder",
				"path": "extending-waypoint/plugin-interfaces/builder"
			},
			{
				"title": "Registry",
				"path": "extending-waypoint/plugin-interfaces/registry"
			},
			{
				"title": "Platform",
				"path": "extending-waypoint/plugin-interfaces/platform"
			},
			{
				"title": "ReleaseManager",
				"path": "extending-waypoint/plugin-interfaces/release-manager"
			},
			{
				"title": "Destroy",
				"path": "extending-waypoint/plugin-interfaces/destroy"
			},
			{
				"title": "Default Parameters",
				"path": "extending-waypoint/plugin-interfaces/default-parameters"
			}
		]
	}
]
