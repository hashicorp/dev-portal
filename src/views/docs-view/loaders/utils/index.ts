/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

export { getNodeFromPath } from './get-node-from-path'
export {
	getPathArraysFromNodes,
	getPathsFromNavData,
} from './get-paths-from-nav-data'
export { normalizeVersion } from './normalize-version'
export { resolveNavData } from './resolve-nav-data'
export { stripVersionFromPathParams } from './strip-version-from-path'
export { validateNavData } from './validate-nav-data'
