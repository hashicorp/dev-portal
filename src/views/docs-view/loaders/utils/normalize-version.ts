/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

export const normalizeVersion = (version: string) => {
	if (version === 'latest') {
		return version
	}
	return version.startsWith('v') ? version : `v${version}`
}
