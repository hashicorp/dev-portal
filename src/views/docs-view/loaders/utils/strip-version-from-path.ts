/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

const REGEX = /^v[0-9]+\.[0-9]+\.(x|[0-9]+)$/i
const TFE_REGEX = /^v[0-9]{6}-[0-9]+$/i

/**
 * Given an array of strings, returns a tuple of
 * - [0]: a version or 'latest'
 * - [1]: an array without the version
 * @param {string[]} pathParams
 * @returns {[string, string[]]} [version, strippedParams]
 */
export const stripVersionFromPathParams = (
	pathParams: string[] = []
): [string, string[]] => {
	const index = pathParams.findIndex((e) => REGEX.test(e) || TFE_REGEX.test(e))
	let version = 'latest'
	let params = [...pathParams]

	// A "version" is expected to only exist at index (0)
	// in a list of Next.js params.
	// If it is elsewhere, it's likely that it should be left alone.

	if (index == 0) {
		version = pathParams[index]
		params = [...params.slice(0, index), ...params.slice(index + 1)]
		return [version, params]
	}

	return [version, params]
}
