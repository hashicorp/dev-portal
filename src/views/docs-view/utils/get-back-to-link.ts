import { SidebarBackToLinkProps } from 'components/sidebar/components/sidebar-back-to-link'
import { ProductData, RootDocsPath } from 'types/products'

/**
 * Generate `backToLinkProps` for the current `rootDocsPath`,
 * for use with `DocsView`.
 *
 * For nested rootDocsPaths, the `backToLinkProps` will
 * lead to the "parent" rootDocsPath if there is one.
 *
 * For other cases, the `backToLinkProps` will lead to the product home page.
 *
 * A rootDocsPath is considered "nested" if its path contains a slash.
 * If we have a "nested" rootDocsPath, we should try to find the parent
 * rootDocsPath. If we can't find a "parent" rootDocsPath,
 * or if this is not a "nested" rootDocsPath, then we should use
 * a back-to "<product> Home" link.
 */
export function getBackToLink(
	currentRootDocsPath: Pick<RootDocsPath, 'name' | 'path'>,
	product: Pick<ProductData, 'name' | 'slug' | 'rootDocsPaths'>
): SidebarBackToLinkProps {
	/**
	 * A rootDocsPath is considered "nested" if it has multiple path parts.
	 */
	const pathParts = currentRootDocsPath.path.split('/')
	const isNestedPath = pathParts.length > 1
	/**
	 * If we have a nested path, we try to find the parent rootDocsPath.
	 * Note that we may not find a match.
	 */
	if (isNestedPath) {
		const parentPath = pathParts.slice(0, pathParts.length - 1).join('/')
		const parentMatches = product.rootDocsPaths.filter((rootDocsPath) => {
			return rootDocsPath.path == parentPath
		})
		if (parentMatches.length > 0) {
			const parentDocsPath = parentMatches[0]
			return {
				text: parentDocsPath.name,
				href: `/${product.slug}/${parentDocsPath.path}`,
			}
		}
	}
	/**
	 * If we can't find a "parent" rootDocsPath,
	 * or if this is not a "nested" rootDocsPath, then we should use
	 * a generic back-to "<product> Home" link.
	 */
	return {
		text: `${product.name} Home`,
		href: `/${product.slug}`,
	}
}
