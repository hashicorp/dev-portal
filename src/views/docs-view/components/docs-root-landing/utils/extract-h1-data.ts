import { Children, ReactChild, ReactElement } from 'react'
import { isPageHeadingChild } from './is-page-heading-child'
import { getElementTextContent } from './get-element-text-content'
import { isElement } from 'react-is'

/**
 * Typeguard that uses `react-is` to ensure a value is a `ReactElement`.
 */
function isReactElement(value): value is ReactElement {
	return isElement(value)
}

/**
 * Iterate over the content children,
 * extracting <h1> text to use as the `heading` for our `LandingHero`,
 * and avoiding duplicative rendering of the <h1>.
 */
export default function extractPageHeading(
	children: ReactChild[]
): { title: string; id: string } | null {
	const headingChild = Children.toArray(children).find(isPageHeadingChild)
	// If we didn't find a matching ReactElement, return `null`
	if (!headingChild || !isReactElement(headingChild)) {
		return null
	}
	// If we did find a match, extract the `title` and `id`, and return them
	return {
		title: getElementTextContent(headingChild),
		id: headingChild.props.id,
	}
}
