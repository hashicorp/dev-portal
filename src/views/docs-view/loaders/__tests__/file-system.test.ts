/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { mocked } from 'jest-mock'
import FileSystemLoader from '../file-system'
import { resolveNavData } from '../utils'
import navData from '../__fixtures__/navData.json'

jest.mock('../utils/resolve-nav-data')
const mockedResolveNavData = mocked(resolveNavData)

const CONTENT_DIR = 'src/views/docs-view/loaders/__fixtures__'

let loader: FileSystemLoader

import * as nextMdxRemote from 'next-mdx-remote/serialize'
const serializeSpy = jest.spyOn(nextMdxRemote, 'serialize')
const mockMdxContentHook = jest.fn()

describe('FileSystemLoader', () => {
	beforeAll(() => {
		loader = new FileSystemLoader({
			navDataFile: 'test',
			localContentDir: CONTENT_DIR,
			product: 'waypoint',
		})
	})

	beforeEach(() => {
		jest.clearAllMocks()
	})

	test('generates path from nav data', async () => {
		mockedResolveNavData.mockImplementation(async () => navData)

		const paths = await loader.loadStaticPaths()

		expect(paths).toMatchSnapshot()
	})

	test('generates static props from mdx file', async () => {
		const props = await loader.loadStaticProps({ params: {} })

		expect(props).toMatchInlineSnapshot(
			{
				mdxSource: { compiledSource: expect.any(String) },
				navData: expect.any(Array),
			},
			`
		Object {
		  "currentPath": "",
		  "frontMatter": Object {
		    "description": "Welcome to the intro guide to Vault! This guide is the best place to start with Vault. We cover what Vault is, what problems it can solve, how it compares to existing software, and contains a quick start for using Vault.",
		    "page_title": "Introduction",
		  },
		  "githubFileUrl": "https://github.com/hashicorp/waypoint/blob/main/website/src/views/docs-view/loaders/__fixtures__/index.mdx",
		  "mdxSource": Object {
		    "compiledSource": Any<String>,
		    "scope": Object {
		      "version": "latest",
		    },
		  },
		  "navData": Any<Array>,
		  "versions": Array [],
		}
	`
		)
	})

	test('uses provided githubFileUrl if provided', async () => {
		const l = new FileSystemLoader({
			navDataFile: 'test',
			localContentDir: CONTENT_DIR,
			product: 'waypoint',
			githubFileUrl(p) {
				return `https://hashicorp.com/${p}`
			},
		})
		const props = await l.loadStaticProps({ params: {} })

		expect(props).toMatchInlineSnapshot(
			{
				mdxSource: { compiledSource: expect.any(String) },
				navData: expect.any(Array),
			},
			`
		Object {
		  "currentPath": "",
		  "frontMatter": Object {
		    "description": "Welcome to the intro guide to Vault! This guide is the best place to start with Vault. We cover what Vault is, what problems it can solve, how it compares to existing software, and contains a quick start for using Vault.",
		    "page_title": "Introduction",
		  },
		  "githubFileUrl": "https://hashicorp.com/src/views/docs-view/loaders/__fixtures__/index.mdx",
		  "mdxSource": Object {
		    "compiledSource": Any<String>,
		    "scope": Object {
		      "version": "latest",
		    },
		  },
		  "navData": Any<Array>,
		  "versions": Array [],
		}
	`
		)
	})

	test('mdxContentHook is called if provided', async () => {
		mockMdxContentHook.mockImplementation((mdxContent, scope) => 'Mock impl')

		const l = new FileSystemLoader({
			navDataFile: 'test',
			localContentDir: CONTENT_DIR,
			product: 'waypoint',
			githubFileUrl(p) {
				return `https://hashicorp.com/${p}`
			},
			mdxContentHook: mockMdxContentHook,
			scope: { version: 'latest' },
		})

		await l.loadStaticProps({ params: {} })

		expect(mockMdxContentHook).toHaveBeenCalledWith(expect.any(String), {
			version: 'latest',
		})
		// assert that `serialize` is called with the result of the hook
		expect(serializeSpy).toHaveBeenCalledWith('Mock impl', expect.any(Object))
	})
})
