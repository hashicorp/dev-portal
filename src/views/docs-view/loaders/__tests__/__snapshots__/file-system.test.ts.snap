// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSystemLoader generates path from nav data 1`] = `
Array [
  Object {
    "params": Object {
      "page": Array [],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
        "main",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
        "configuration",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
        "build-interface",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
        "compiling",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
        "example-application",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "creating-plugins",
        "testing",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "main-func",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "passing-values",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "authenticator",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "configurable",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "configurable-notify",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "builder",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "registry",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "platform",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "release-manager",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "destroy",
      ],
    },
  },
  Object {
    "params": Object {
      "page": Array [
        "extending-waypoint",
        "plugin-interfaces",
        "default-parameters",
      ],
    },
  },
]
`;
